designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,613 @@
+
+```js
+// eslint.config.js
+import js from "@eslint/js";
+
+export default [
+    {
+        ...js.configs.recommended,
+        files: ["**/src/safe/*.js"]
+    }
+];
+```
+
+For arrays you do this:
+
+```js
+// eslint.config.js
+import exampleConfigs from "eslint-config-example";
+
+export default [
+    ...exampleConfigs.map(config => ({
+        ...config,
+        files: ["**/src/safe/*.js"]
+    })),
+
+    // your modifications
+    {
+        rules: {
+            "no-unused-vars": "warn"
+        }
+    }
+];
+```
+
+In both cases, users need to use the spread syntax, and in both cases, just glancing at the config object doesn't really tell you what exactly is happening. It's a lot of syntax for what used to be a single statement in eslintrc.
+
+## Goals
+
+There are two goals with this design:
+
+1. **Make it easier to configure ESLint.** For users, we want to reduce unnecessarily boilerplate and guesswork as much as possible.
+1. **Encourage plugins to use `configs`.** For plugin authors, we want to encourage the use of a consistent entrypoint, `configs`, where users can find predefined configurations. 
+
+## Detailed Design
+
+Design Summary:
+
+1. Allow arrays in config arrays (in addition to objects)
+1. Introduce an `extends` key in config objects
+
+### Allow Arrays in Config Arrays
+
+This was part of the original RFC but was disabled over concerns about the complexity involved with importing configs from packages and having no idea whether it was an object, an array, or an array with a mix of objects and arrays. This mattered most when users needed to use `map()` to apply different `files` or `ignores` to configs, but with `extends`, that concern goes away.
+
+Support for nested arrays is already in `ConfigArray`, it's just disabled.
+
+### Introduce an `extends` Key
+
+The `extends` key is an array that may contain other configurations. When used, ESLint will expand the config object into multiple config objects in order to create a flat structure to evaluate. 
+
+#### Extending Objects
+
+You can pass one or more objects in `extends`, like this:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+
+    {
+        files: ["**/src/*.js"],
+        extends: [js.configs.recommended, example.configs.recommended]
+    }
+
+];
+```
+
+Assuming these config objects do not contain `files` or `ignores`, ESLint will convert this config into the following:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+    {
+        ...js.configs.recommended
+        files: ["**/src/*.js"],
+    },
+    {
+        ...example.configs.recommended
+        files: ["**/src/*.js"],
+    },
+];
+```
+
+If the objects in `extends` contain `files`, then ESLint will intersect those values (AND operation); if the object in `extends` contains `ignores`, then ESLint will merge those values (OR operation). For example:
```

Comment:
I'd like to make one more plea to rethink this behavior to be overriding instead of intersecting. I'm currently trying to set up a fairly out-of-the-box project with vue and TS for my day job.

So, I have this setup
```js
import pluginVue from 'eslint-plugin-vue';
import tseslint from 'typescript-eslint';
import eslint from '@eslint/js';

export default [
  {
    ignores: ['node_modules', 'dist', 'dist-ssr'],
  },
  eslint.configs.recommended,
  tseslint.configs.eslintRecommended,
  tseslint.configs.recommended,
  {
    languageOptions: {
      parserOptions: {
        projectService: true,
        tsconfigRootDir: import.meta.filename,
      },
    },
  },
  {
    files: ['**/*.vue'],
    languageOptions: {
      parserOptions: {
        parser: tseslint.parser,
        extraFileExtensions: ['.vue'],
      },
    },
  },
  pluginVue.configs['flat/recommended'],
].flat(Infinity);
```

Did you catch the error? The base eslint config (`eslint.configs.recommended`) applies to all files regardless of file extension, and the `tseslint.configs.eslintRecommended` config applies to JS/TS-like files only. So I get false positives like `no-undef` and others in my *.vue TS code, since I haven't applied `tseslint.configs.eslintRecommended` to it. 

As a user, I don't know or much care which config is in the right about the file extensions to which it applies. But, I want to be able to fix this issue easily enough. `tseslint.configs.eslintRecommended` doesn't apply to vue files? No problem; I'll chuck it in as an extends to my vue-specific config. With `tseslint.config()` that looks like:

```js
import pluginVue from 'eslint-plugin-vue';
import tseslint from 'typescript-eslint';
import eslint from '@eslint/js';

export default tseslint.config(
  {
    ignores: ['node_modules', 'dist', 'dist-ssr'],
  },
  eslint.configs.recommended,
  tseslint.configs.eslintRecommended,
  tseslint.configs.recommended,
  {
    languageOptions: {
      parserOptions: {
        projectService: true,
        tsconfigRootDir: import.meta.filename,
      },
    },
  },
  {
    files: ['**/*.vue'],
    extends: [tseslint.configs.eslintRecommended],
    languageOptions: {
      parserOptions: {
        parser: tseslint.parser,
        extraFileExtensions: ['.vue'],
      },
    },
  },
  pluginVue.configs['flat/recommended'],
);
```

Done! And intuitive IMO.

What shall I do if `extends` is an intersection? Maybe this?

```js
import pluginVue from 'eslint-plugin-vue';
import tseslint from 'typescript-eslint';
import eslint from '@eslint/js';

export default [
  {
    ignores: ['node_modules', 'dist', 'dist-ssr'],
  },
  eslint.configs.recommended,
  tseslint.configs.eslintRecommended,
  tseslint.configs.recommended,
  {
    languageOptions: {
      parserOptions: {
        projectService: true,
        tsconfigRootDir: import.meta.filename,
      },
    },
  },
  {
    files: ['**/*.vue'],
    languageOptions: {
      parserOptions: {
        parser: tseslint.parser,
        extraFileExtensions: ['.vue'],
      },
    },
    extends:
      (() => {
        const maybeMultipleConfigs = structuredClone([tseslint.configs.eslintRecommended]);
        const flattenedConfigs = maybeMultipleConfigs.flat(Infinity);
        for (const config of flattenedConfigs) {
          delete config.files;
        }
        return flattenedConfigs;
      })(),
  },
  pluginVue.configs['flat/recommended'],
];
```

How will a user possibly _expand_ the `include`s of an arbitrarily deeply nested config array if they only operation they have access to is intersection? I don't see how it can be done without nontrivial amounts of user-defined JS... which is the express purpose of this rfc to eliminate.

Reply:
I feel like the relevant points were covered in [this thread](https://github.com/eslint/rfcs/pull/126#discussion_r1843063505) already, so I don't want to duplicate here all of that discussion here.

Intersection is what eslintrc did, and I think it still makes the most sense for the majority of users. Overriding means we lose the ability to extend a config array in any meaningful way -- there's really not a great reason to provide an array if you can't limit each object to a subset of files, and overwriting `files` for every config object in an extended array is unlikely to produce the desired result.